overlays/haskell.nix

```diff
@@ -551,10 +552,50 @@ final: prev: {
                 ) rawProject.hsPkgs
                 // {
                   # These are functions not packages
-                  inherit (rawProject.hsPkgs) shellFor makeConfigFiles ghcWithHoogle ghcWithPackages;
+                  inherit (rawProject.hsPkgs) makeConfigFiles ghcWithHoogle ghcWithPackages;
+                  shellFor = a: __trace "project.hsPkgs.shellFor has been deprecated in favour of project.shellFor"
+                    (rawProject.hsPkgs.shellFor a);
               });
 
             projectCoverageReport = haskellLib.projectCoverageReport (map (pkg: pkg.coverageReport) (final.lib.attrValues (haskellLib.selectProjectPackages hsPkgs)));
+
+            cross = (final.lib.mapAttrs (_: pkgs:
+                rawProject.projectFunction pkgs.haskell-nix
+              ) final.pkgsCross) // { recurseForDerivations = false; };
+
+            # Add support for passing in `cross` argument
+            shellFor = shellArgs:
+              if shellArgs ? cross
+                then
+                  let
+                    otherArgs = builtins.removeAttrs shellArgs [ "cross" ];
+                    crossShells = builtins.map (pkgs:
+                      (rawProject.projectFunction pkgs.haskell-nix).shellFor (otherArgs // {
+                        buildInputs = [];
+                        nativeBuildInputs = [];
+                      }))
+                      (shellArgs.cross final.pkgsCross);
+                  in project.hsPkgs.shellFor (otherArgs // {
```

Comment:
I feel like some of this should be in the shellFor builder, perhaps?

Reply:
I had a go at moving some of it in and it does seem a lot nice (in particular putting the cabal wrapper in there is much nicer).  Added an `otherShells` arg that is an easy way of including the `buildInputs` and `nativeBuildInputs` of other shells.

I left the stuff relating to calling `crossPlatforms` and using `projectFunction` in the haskell overlay.